MiauAPI/Services/ProductImageService.cs

```diff
@@ -0,0 +1,138 @@
+    /// <param name="location">The URL of the new resource or the content of the Location header.</param>
+    /// <returns>The result of the operation.</returns>
+    /// <exception cref="ArgumentException">Occurs when <paramref name="location"/> is <see langword="null"/> or empty.</exception>
+    public async Task<ActionResult<OneOf<ProductImageResponse, ErrorResponse>>> CreatedProductImageAsync(ProductImageRequest request, string location)
+    {
+        if (string.IsNullOrWhiteSpace(location))
+            throw new ArgumentException("Location cannot be null or empty.", nameof(location));
+
+        // Checks if request contains valid data
+        if (!_validator.IsRequestValid(request, out var errorMessages))
+            return new BadRequestObjectResult(new ErrorResponse(errorMessages.ToArray()));
+
+        // Checks the ProductId
+        var dbProduct = await _db.Products.FirstAsync(x => x.Id == request.ProductId);
+        if (dbProduct == null)
+        {
+            new NotFoundObjectResult(new ErrorResponse($"No product with the Id = {request.ProductId} was found"));
+        }
+
+        //Creates the path for the Product Image
+        var path = $"Data/{request.ProductId}/images";
+        if ((!Directory.Exists(path)))
+        {
+            Directory.CreateDirectory(path);
+        }
+        var filename = request.ImagePath.FileName;
+        using (var fileStream = new FileStream(Path.Combine(path, filename), FileMode.Create))
+        {
+            await request.ImagePath.CopyToAsync(fileStream);
+        }
+
+        // Create the database product image
+        var dbProductImage = new ProductImageEntity()
+        {
+            Product = dbProduct!,
+            FileUrl = $"images/{request.ProductId}/" + filename
+        };
+
+        await _db.ProductImages.AddAsync(dbProductImage);
+        await _db.SaveChangesAsync();
+
+        // TODO: handle authentication properly
+        return new CreatedResult(location, new ProductImageResponse(dbProductImage.Id));
+    }
+
+    /// <summary>
+    /// Returns a list of product images.
+    /// </summary>
+    /// <returns>The result of the operation.</returns>
+    public async Task<ActionResult<OneOf<GetProductImageResponse, ErrorResponse>>> GetProductImageAsync(ProductImageParameters productImageParameters)
+    {
+
+        var errorMessages = Enumerable.Empty<string>();
+        var dbProductImages = _db.ProductImages.Select(p => new ProductImageObject
+        {
+            Id = p.Id,
+            ProductId = p.Product.Id,
+            ImagePath = p.FileUrl
+        });
+
+        if (productImageParameters.ProductId != 0)
+        {
+            dbProductImages = dbProductImages.Where(p => p.ProductId == productImageParameters.ProductId);
+        }
+
+        if (productImageParameters.Id != 0)
+        {
+            dbProductImages = dbProductImages.Where(p => p.Id == productImageParameters.Id);
+        }
```

Comment:
Existe alguma vantagem de se construir a query caso a requisição não especifique o ID do produto?
Eu provavelmente só retornaria BadRequest se qualquer um dos ids for igual a 0.

Reply:
Matheus, não entendi. 

Eu tinha pensado em 3 cenários:
- O Admin querer acessar todas as imagens 
- O Admin querer acessar todas as imagens de um mesmo produto (necessário o `ProductId`)
- O Admin querer acessar uma única imagem  (necessário o `Id` da imagem).

Você diz algo tipo assim?

```cs
if (productImageParameters.ProductId == 0 || productImageParameters.Id == 0)
{
        return new BadRequestResult();
}
dbProductImages = dbProductImages.Where(p => p.ProductId == productImageParameters.ProductId)
                                 .Where(p => p.Id == productImageParameters.Id);
```

Reply:
Ah, faz sentido.
Pode deixar do jeito que está então.